### PR TITLE
Test templates fixes

### LINKED ISSUE #3981: Add code in test project templates to configure HttpClient resiliency
We should consider updating the Aspire integration test project templates to have code that adds `HttpClient` resiliency now that it was removed from `DistributedApplicationTestBuilder`, e.g.:

```csharp
builder.Services.ConfigureHttpClientDefaults(clientBuilder =>
{
    clientBuilder.AddStandardResilienceHandler();
});
```

### LINKED ISSUE #4024: Starter template with `.9` in the name generates a broken project
Project with a `.9` in it's name (period followed by a number):

1. `dotnet new aspire-starter -o "starter v.9"`
2.  `dotnet run` for the AppHost project:

```
/private/tmp/test/starter v.9/starter v.9.AppHost/Program.cs(3,46):
error CS0234: The type or namespace name 'starter_v__9_ApiService' does not exist in the namespace 'Projects' (are you missing an assembly reference?) [/private/tmp/test/starter v.9/starter v.9.AppHost/starter v.9.AppHost.csproj]
/private/tmp/test/starter v.9/starter v.9.AppHost/Program.cs(5,29):
error CS0234: The type or namespace name 'starter_v__9_Web' does not exist in the namespace 'Projects' (are you missing an assembly reference?) [/private/tmp/test/starter v.9/starter v.9.AppHost/starter v.9.AppHost.csproj]

The build failed. Fix the build errors and run again.
```

The generated classes: `public class starter_v_9_ApiService`, and `public class starter_v_9_Web`.

cc @timheuer @DamianEdwards @joperezr

### LINKED ISSUE #4025: Starter template with name `starter xyz.foo-bar` generates a broken Test project
1. `dotnet new aspire-starter -t mstest -o "starter xyz.foo-bar"`
2.  `dotnet run` for `AppHost` works fine
3. `dotnet test` for the `Tests` projects fails to build:

`/private/tmp/test/starter xyz.foo-bar/starter xyz.foo-bar.Tests/WebTests.cs(12,87): error CS0234: The type or namespace name 'starter_xyz' does not exist in the namespace 'Projects' (are you missing an assembly reference?) [/private/tmp/test/starter xyz.foo-bar/starter xyz.foo-bar.Tests/starter xyz.foo-bar.Tests.csproj]`

`WebTests.cs` line 12: `var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.starter_xyz.foo_bar_AppHost>();`

Generated class name in `obj/Debug/net8.0/Aspire/references/_AppHost.ProjectMetadata.g.cs`:
`public class starter_xyz_foo_bar_Tests`

The generated class name does not seem to follow the same rules as the template name generation.

cc @timheuer @DamianEdwards @joperezr

### LINKED ISSUE #4056: `aspire-mstest`, and `aspire-nunit` templates are broken
Steps to reproduce:

1. `dotnet new aspire -o asp0`
2. `cd asp0; dotnet new aspire-nunit -o testProject`

It is trying to restore `Aspire.Tests1.csproj` instead of `testProject.csproj`.

Screenshot 2024-05-01 at 19 09 48

Fix:
```diff
diff --git a/src/Aspire.ProjectTemplates/templates/aspire-mstest/.template.config/template.json b/src/Aspire.ProjectTemplates/templates/aspire-mstest/.template.config/template.json
index 516fd9c49..31c19eb8b 100644
--- a/src/Aspire.ProjectTemplates/templates/aspire-mstest/.template.config/template.json
+++ b/src/Aspire.ProjectTemplates/templates/aspire-mstest/.template.config/template.json
@@ -58,7 +58,7 @@
   },
   "primaryOutputs": [
     {
-      "path": "Aspire.Tests1.csproj"
+      "path": "Aspire.Tests.1.csproj"
     }
   ],
   "postActions": [
diff --git a/src/Aspire.ProjectTemplates/templates/aspire-nunit/.template.config/template.json b/src/Aspire.ProjectTemplates/templates/aspire-nunit/.template.config/template.json
index 7574e8980..a2a43d986 100644
--- a/src/Aspire.ProjectTemplates/templates/aspire-nunit/.template.config/template.json
+++ b/src/Aspire.ProjectTemplates/templates/aspire-nunit/.template.config/template.json
@@ -58,7 +58,7 @@
   },
   "primaryOutputs": [
     {
-      "path": "Aspire.Tests1.csproj"
+      "path": "Aspire.Tests.1.csproj"
     }
   ],
   "postActions": [
```

cc @DamianEdwards feel free to include this in your PR, or I can open a separate one.
cc @joperezr

### PR DESCRIPTION
**Pull request by DamianEdwards**
~This appears to fix #4025~
~For #4024 I think the targets change I make here fixes that but I haven't fully verified that as it's a bit trickier to do so locally.~

Fixes #4025
Fixes #4024
Fixes #4056
Fixes #3981

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4043)

### CODE REVIEW

**Comment by radical** [comment]
I tried:
```diff
diff --git a/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.Tests/WebTests.cs b/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.Tests/WebTests.cs
index a0a114dde..bf183ff63 100644
--- a/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.Tests/WebTests.cs
+++ b/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.Tests/WebTests.cs
@@ -17,7 +17,7 @@ public class WebTests
     public async Task GetWebResourceRootReturnsOkStatusCode()
     {
         // Arrange
-        var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.AspireStarterApplication._1_AppHost>();
+        var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.AspireStarterApplication__1_AppHost>();
         await using var app = await appHost.BuildAsync();
         await app.StartAsync();
```

.. essentially use to the `className` variation which would always change `.` to `_`, and thus be safe.

.. which fixes #4025 . For #4024, the targets need to be updated to handle the case of `.N` and instead of replacing that with `_N` use `__N`.

But there is yet another case, `aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞` which fails with:
`/private/var/folders/dj/z2lsb9l9087356sgyzrlklvr0000gn/T/testroot/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost.csproj(13,51): error MSB4025: The project file could not be loaded. An error occurred while parsing EntityName. Line 13, position 51.`

.. and the relevant lines:
```xml
    <ProjectReference Include="../aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.ApiService/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.ApiService.csproj" />
    <ProjectReference Include="../aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.Web/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.Web.csproj" />
```

.. `&` is causing the invalid xml. I got this case from @joeloff .

**Comment by DamianEdwards** [comment]
Hmm so the template engine's default form transform doesn't auto-XML-encode in .csproj files? Wouldn't all other templates that have project references based on the name have the same issue too then?

**Comment by radical** [comment]
> Hmm so the template engine's default form transform doesn't auto-XML-encode in .csproj files? Wouldn't all other templates that have project references based on the name have the same issue too then?

Yeah. I can't even create a `blazor` project with that name.

```
 $ dotnet new blazor -o "aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞"
The template "Blazor Web App" was created successfully.
This template contains technologies from parties other than Microsoft, see https://aka.ms/aspnetcore/8.0-third-party-notices for details.

Processing post-creation actions...
Restoring /private/tmp/test/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.csproj:
/private/tmp/test/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞/aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.csproj(7,36): error MSB4025: The project file could not be loaded. An error occurred while parsing EntityName. Line 7, position 36.
Restore failed.
Post action failed.
Manual instructions: Run 'dotnet restore'
```

@joeloff is there an existing way to get a xml-encoded transformation of the string in the template?

If I remove the `&` from the name then I can create `aspire-starter` project but it fails to build:
```
/private/tmp/test/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost/obj/Debug/net8.0/Aspire/references/aspire_龦唉丂荳◎℉㐁ᠭᠤ_སྲིད_خەل_ꌠ_ᥕ᧞_Web.ProjectMetadata.g.cs(8,25): error CS151
4: { expected [/private/tmp/test/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost.csproj]
/private/tmp/test/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost/obj/Debug/net8.0/Aspire/references/_AppHost.ProjectMetadata.g.cs(8,25): error CS1514: { expected [/private/tmp/tes
t/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost/aspire 龦唉丂荳◎℉㐁ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost.csproj]
```

And this is because the code we generate is:
```csharp
public class aspire_    UNs%    !94-$_frQ_.D_ _U_AppHost
{
    private aspire_     UNs%    !94-$_frQ_.D_ _U_AppHost() { }
    public static string ProjectPath => """/private/tmp/test/aspire     UNs%    !94-$ frQ .D   U/aspire         UNs%    !94-$ frQ .D   U.AppHost""";
}
```

.. and this is what the template engine used:
```csharp
var apiService = builder.AddProject<Projects.aspire_龦唉丂荳___㐁ᠭᠤ_སྲིད_خەل_ꌠ_ᥕ__ApiService>("apiservice");
```

**Comment by radical** [comment]
`xmlEncode` transform:

https://github.com/dotnet/templating/blob/a325ffa18edd1590f9b340cf83d51d8eb567ebdc/docs/Value-Forms.md?plain=1#L12-L31

In my source, the value `"Hi there"` should be replaced by a user supplied value, but, `&quot;Hi there&quot;` should be as well. The user value should be XML encoded for the latter and not for the former. The below snippet from template.json shows how to accomplish this.

```json
{
  ...
  "symbols": {
    "example": {
      "type": "parameter",
      "dataType": "string",
      "replaces": "\"Hi there\"",
      "forms": {
        "global": [ "encode" ]
      }
    }
  },
  "forms": {
    "encode": {
      "identifier": "xmlEncode"
    }
  }
}
```

**Comment by radical** [comment]
```diff
diff --git a/src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json b/src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json
index 652ddb0dd..67d70c756 100644
--- a/src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json
+++ b/src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json
@@ -344,6 +344,13 @@
       "datatype": "bool",
       "defaultValue": "false",
       "description": "Whether to turn off HTTPS."
+    },
+    "XmlEncodedProjectName": {
+      "type": "derived",
+      "datatype": "string",
+      "replaces": "XmlEncodedProjectName",
+      "valueSource": "name",
+      "valueTransform": "xmlEncode"
     }
   },
   "primaryOutputs": [
diff --git a/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.ApiService/AspireStarterApplication.1.ApiService.csproj b/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireS>
index f7812db4b..3b56d6e4b 100644
--- a/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.ApiService/AspireStarterApplication.1.ApiService.csproj
+++ b/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.ApiService/AspireStarterApplication.1.ApiService.csproj
@@ -7,7 +7,7 @@
   </PropertyGroup>

   <ItemGroup>
-    <ProjectReference Include="..\AspireStarterApplication.1.ServiceDefaults\AspireStarterApplication.1.ServiceDefaults.csproj" />
+    <ProjectReference Include="..\XmlEncodedProjectName.ServiceDefaults\XmlEncodedProjectName.ServiceDefaults.csproj" />
   </ItemGroup>

 </Project>
diff --git a/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.AppHost/AspireStarterApplication.1.AppHost.csproj b/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarter>
index 4652b02ef..40434d256 100644
--- a/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.AppHost/AspireStarterApplication.1.AppHost.csproj
+++ b/src/Aspire.ProjectTemplates/templates/aspire-starter/AspireStarterApplication.1.AppHost/AspireStarterApplication.1.AppHost.csproj
@@ -10,8 +10,8 @@
   </PropertyGroup>

   <ItemGroup>
-    <ProjectReference Include="..\AspireStarterApplication.1.ApiService\AspireStarterApplication.1.ApiService.csproj" />
-    <ProjectReference Include="..\AspireStarterApplication.1.Web\AspireStarterApplication.1.Web.csproj" />
+    <ProjectReference Include="..\XmlEncodedProjectName.ApiService\XmlEncodedProjectName.ApiService.csproj" />
+    <ProjectReference Include="..\XmlEncodedProjectName.Web\XmlEncodedProjectName.Web.csproj" />
   </ItemGroup>

   <ItemGroup>
```

.. this results in a fixed: `<ProjectReference Include="../aspire 龦唉丂荳◎℉㐁&amp;ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.ServiceDefaults/aspire 龦唉丂荳◎℉㐁&amp;ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.ServiceDefaults.csproj" />`

**Comment by DamianEdwards** [comment]
@radical I tried applying your suggested change locally and ran into issues with the .sln file not working apparently. All the paths and names look correct, but I get the sense this might never have worked properly in any template judging from how much of the logging doesn't even seem to support the unicode chars:

```
dotnet new aspire-starter -t MSTest -o "aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞"
The template ".NET Aspire Starter Application" was created successfully.
This template contains technologies from parties other than Microsoft, see https://aka.ms/dotnet/aspire/8.0-third-party-notices for details.

Processing post-creation actions...
Restoring D:\src\local\Aspire\aspire ????????&?? ???? ??? ? ??\aspire ????????&?? ???? ??? ? ??.sln:
C:\Program Files\dotnet\sdk\8.0.300-preview.24203.14\NuGet.targets(414,5): error MSB3202: The project file "D:\src\local\Aspir
e\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire é¾¦å”‰ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.AppHost\aspire é¾¦å”‰
ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.AppHost.csproj" was not found. [D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭ
ᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.sln]
C:\Program Files\dotnet\sdk\8.0.300-preview.24203.14\NuGet.targets(414,5): error MSB3202: The project file "D:\src\local\Aspir
e\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire é¾¦å”‰ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.ServiceDefaults\aspir
e é¾¦å”‰ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.ServiceDefaults.csproj" was not found. [D:\src\local\Aspire\a
spire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.sln]
C:\Program Files\dotnet\sdk\8.0.300-preview.24203.14\NuGet.targets(414,5): error MSB3202: The project file "D:\src\local\Aspir
e\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire é¾¦å”‰ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.ApiService\aspire é¾¦
å”‰ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.ApiService.csproj" was not found. [D:\src\local\Aspire\aspire 龦唉丂荳
◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.sln]
C:\Program Files\dotnet\sdk\8.0.300-preview.24203.14\NuGet.targets(414,5): error MSB3202: The project file "D:\src\local\Aspir
e\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire é¾¦å”‰ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.Tests\aspire é¾¦å”‰ä¸
‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.Tests.csproj" was not found. [D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲ
ིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.sln]
C:\Program Files\dotnet\sdk\8.0.300-preview.24203.14\NuGet.targets(414,5): error MSB3202: The project file "D:\src\local\Aspir
e\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire é¾¦å”‰ä¸‚è³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.Web\aspire é¾¦å”‰ä¸‚è
³â—Žâ„‰î€¹ã&á ­á ¤ à½¦à¾²à½²à½‘ Ø®Û•Ù„ êŒ  á¥•á§ž.Web.csproj" was not found. [D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خ
ەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.sln]
Restore failed.
Restoring D:\src\local\Aspire\aspire ????????&?? ???? ??? ? ??\aspire ????????&?? ???? ??? ? ??.AppHost\aspire ????????&?? ???? ??? ? ??.AppHost.csproj:
  Determining projects to restore...
  Restored D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.ApiService\aspire 龦唉丂荳◎℉㐁&ᠭᠤ
   སྲིད خەل ꌠ ᥕ᧞.ApiService.csproj (in 490 ms).
  Restored D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.ServiceDefaults\aspire 龦唉丂荳◎℉
  㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.ServiceDefaults.csproj (in 490 ms).
  Restored D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.Web\aspire 龦唉丂 荳◎℉㐁&ᠭᠤ སྲིད خ
  ەل ꌠ ᥕ᧞.Web.csproj (in 490 ms).
  Restored D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.AppHost\aspire 龦 唉丂荳◎℉㐁&ᠭᠤ སྲ
  ིད خەل ꌠ ᥕ᧞.AppHost.csproj (in 490 ms).
Restore succeeded.
Restoring D:\src\local\Aspire\aspire ????????&?? ???? ??? ? ??\aspire ????????&?? ???? ??? ? ??.ServiceDefaults\aspire ????????&?? ???? ??? ? ??.ServiceDefaults.csproj:
  Determining projects to restore...
  All projects are up-to-date for restore.
Restore succeeded.
Restoring D:\src\local\Aspire\aspire ????????&?? ???? ??? ? ??\aspire ????????&?? ???? ??? ? ??.ApiService\aspire ????????&?? ???? ??? ? ??.ApiService.csproj:
  Determining projects to restore...
  All projects are up-to-date for restore.
Restore succeeded.
Restoring D:\src\local\Aspire\aspire ????????&?? ???? ??? ? ??\aspire ????????&?? ???? ??? ? ??.Web\aspire ????????&?? ???? ??? ? ??.Web.csproj:
  Determining projects to restore...
  All projects are up-to-date for restore.
Restore succeeded.
Restoring D:\src\local\Aspire\aspire ????????&?? ???? ??? ? ??\aspire ????????&?? ???? ??? ? ??.Tests\aspire ????????&?? ???? ??? ? ??.Tests.csproj:
  Determining projects to restore...
  Restored D:\src\local\Aspire\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞\aspire 龦唉丂荳◎℉㐁&ᠭᠤ སྲིད خەل ꌠ ᥕ᧞.Tests\aspire 龦唉 丂荳◎℉㐁&ᠭᠤ སྲིད
   خەل ꌠ ᥕ᧞.Tests.csproj (in 152 ms).
  4 of 5 projects are up-to-date for restore.
Restore succeeded.
Post action failed.
Manual instructions: Run 'dotnet restore'
```

**Comment by radical** [comment]
Maybe we can skip the case where the name has characters needing to be xml encoded for now. But other than that we should get unicode chars working, right?

**Comment by DamianEdwards** [comment]
@radical I'm not sure we fully understand exactly what's supported though, e.g. that example you gave results in C# errors:

image

**Comment by radical** [comment]
Starter template works with this on macOS: https://gist.github.com/radical/840d12dcf55ca52d867ed8cb2aeafa34

- stash the template-generated classname in a property in the .csproj
- Use that stashed value in the classname generated in Hosting.targets
- And remove `Encoding="unicode"` from `WriteLinesToFile` to let it default to `utf-8`

**Comment by radical** [comment]
It has scope for cleanup of course :D

**Comment by DamianEdwards** [comment]
@radical are you saying it works with something like `dotnet new aspire-starter -t mstest -o "aspire 龦唉丂荳◎"`? Because I'm still seeing errors including C# and ,sln file parsing. The latest changes I added go with the other suggestion you made regarding have a regex symbol in the .template.json that matches what our targets do for the generated class name. That seems to fix #4024 and #4025 but these extended character examples all seem to result in issues that aren't with our template.

**Comment by radical** [comment]
In my tests I didn't use `-o "aspire 龦唉丂荳◎"`, instead I'm creating a `aspire 龦唉丂荳◎` directory and then running `dotnet new aspire-starter` in that.

https://gist.githubusercontent.com/radical/b0b191a6255f42445c382167534d74ea/raw/e3b904dd6b3ca8482bfc53f6fc6e14b1ad2b26f8/aspire-starter-with-unicode-chars-in-name

But I did try `-o "aspire 龦唉丂荳◎"` and that works fine too. Also, my patch has changed a bit to handle the case where the project doesn't have the new property.

https://gist.github.com/radical/fa68b403ddfa165b725a1fc42901ab69

(Also, the new property is something that we can probably remove by doing similar transformation as yours for the project name)

**Comment by DamianEdwards** [comment]
Hmm I'm trying to figure out what I'm missing in my changes so far then. We definitely don't want a property in the generated project file, so any idea what I should change to get it working?

**Comment by DamianEdwards** [comment]
@radical can you share all the files from the project created with your changes? I'm a bit confused how yours is working as mine seems to be producing the files I expect (names and content) but I then hit C# errors and the solution saying the projects don't exist, but if I copy the path of the projects in the .sln file and do a `dir` on them, they're there 🤷

**Comment by radical** [comment]
[aspire-project-with-unicode-chars-in-name.zip](https://github.com/dotnet/aspire/files/15182089/aspire-project-with-unicode-chars-in-name.zip)

**Comment by radical** [comment]
I haven't tried this on windows yet, btw.

**Comment by DamianEdwards** [comment]
hmm i can't open that zip, WIndows says it's invalid :sad:

**Comment by DamianEdwards** [comment]
ok yeah, that fails on Windows :smile:

**Comment by DamianEdwards** [comment]
@radical Based on what we just figured out I'm marking this ready for review as it does fix the issues logged and it seems Windows doesn't support project names with unicode chars anyway.

**Comment by radical** [comment]
> ok yeah, that fails on Windows 😄

interesting! fyi, I'm pushing my changes to https://github.com/dotnet/aspire/pull/3270 but it has workload tests also. The templates relevant changes are in `src/Aspire.ProjectTemplates/`, and  `src/Aspire.Hosting.AppHost/`, which you can overwrite with yours to run the tests.

> @radical Based on what we just figured out I'm marking this ready for review as it does fix the issues logged and it seems Windows doesn't support project names with unicode chars anyway.

Fair enough. I'll try on windows too, and see if anything is doable but all that can be in separate PR anyway.

File: src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json [hunk 1]
```diff
@@ -344,27 +344,57 @@
...
+            "replacement": "_"
+          },
+          {
+            "regex": "\\-",
+            "replacement": "_"
+          },
+          {
+            "regex": "[ ]",
+            "replacement": "_"
+          }
```
Comment (radical) [inline src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json:367]: Can this be one regex `"[\\.\\- ]"`?

Comment (DamianEdwards) [reply src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json:367]: Let me try

**Comment by radical** [comment]
> ok yeah, that fails on Windows 😄

Not for this PR, but I think `chcp.com 65001` would be needed in `cmd` to build the project, or powershell.

**Comment by radical** [comment]
I tried the changes here with the workload tests, and they passed.

File: src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json [hunk 2]
```diff
@@ -15,7 +15,7 @@
   "defaultName": "AspireApp",
   "description": "A project template for creating a .NET Aspire app with a Blazor web frontend and web API backend service, optionally using Redis for caching.",
   "shortName": "aspire-starter",
-  "sourceName": "AspireStarterApplication.1",
+  "sourceName": "Aspire-StarterApplication.1",
```
[the remark this answers is not in the document]
Comment (DamianEdwards) [reply src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json:18]: It helps to create some separation between "Aspire" and "StarterApplication", and adds another character that is different in the different forms, i.e. `-` becomes `_` in namespace and class name form, while `.` remains `.` in namespaces and becomes `_` in class names. Not strictly necessary but I think it makes it a bit clearer.

File: src/Aspire.ProjectTemplates/templates/aspire-starter/Aspire-StarterApplication.1.Tests/WebTests.cs [hunk 3]
```diff
@@ -1,6 +1,6 @@
 using System.Net;
 
-namespace AspireStarterApplication._1.Tests;
```
Comment (radical) [inline src/Aspire.ProjectTemplates/templates/aspire-starter/Aspire-StarterApplication.1.Tests/WebTests.cs:3]: Also add `using Microsoft.Extensions.DependencyInjection;` to the various places `ConfigureHttpClientDefaults` is being used now, to fix
```
/private/var/folders/dj/z2lsb9l9087356sgyzrlklvr0000gn/T/testroot/starter test.Release-xunit_net_sp4yrph3_r2i/starter test.Release-xunit_net_sp4yrph3_r2i.Tests/WebTests.cs(12,26): 
error CS1061: 'IServiceCollection' does not contain a definition for 'ConfigureHttpClientDefaults' and no accessible extension method 'ConfigureHttpClientDefaults' accepting a first argument of type 'IServiceCollection' could be found (are you missing a using directive or an assembly reference?) 
```

File: src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json [hunk 4]
```diff
@@ -344,27 +344,49 @@
...
+    "GeneratedClassNamePrefix": {
+      "type": "generated",
+      "generator": "regex",
+      "datatype": "string",
+      "replaces": "GeneratedClassNamePrefix",
+      "parameters": {
+        "source": "name",
+        "steps": [
+          {
+            "regex": "[\\.\\- ]",
```
Comment (radical) [inline src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json:357]: The template engine uses this for the class name: 
https://github.com/dotnet/templating/blob/a325ffa18edd1590f9b340cf83d51d8eb567ebdc/src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/DefaultSafeNameValueFormFactory.cs#L19

.. which seems to be better as it catches more cases - `(((?<=\.)|^)(?=\d)|\W)`.

Comment (radical) [reply src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json:357]: This patch fixes it for both template generated names, and our own - https://gist.github.com/radical/5c746337ecf8136fa240f4fd1ea1086f 

Tested with `"aspire & foo-bar"`.

Comment (DamianEdwards) [reply src/Aspire.ProjectTemplates/templates/aspire-starter/.template.config/template.json:357]: It needs to match what we do in our targets though

**Review by radical — APPROVED** [review]

**Comment by radical** [comment]
Thank you for being patient, @DamianEdwards!

**Pull request merged**